infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -618,27 +709,35 @@ public void reset() {
 
         try {
             long committedTail = getCommittedTail();
-            long latestSegment = getSegmentId(getLogTail());
-            long committedTailSegment = getSegmentId(committedTail);
-            for (long currSegmentId = committedTailSegment; currSegmentId <= latestSegment; currSegmentId++) {
-                // Close segments before deleting their corresponding log files
-                String segmentFilePath;
-                try (Segment sh = getSegmentHandleForSegment(currSegmentId)) {
-                    openSegments.remove(sh.id);
-                    segmentFilePath = sh.segmentFilePath;
-                }
+            long globalTail = getLogTail();
+            long newTailAddress = StreamLogDataStore.ZERO_ADDRESS;
 
-                deleteFilesMatchingFilter(file -> file.getAbsolutePath().equals(segmentFilePath));
-            }
+            if (committedTail != globalTail) {
+                // Delete all segments up to the committed tail's segment
+                long latestSegment = getSegmentId(globalTail);
+                long committedTailSegment = getSegmentId(committedTail);
+
+                for (long currSegmentId = committedTailSegment; currSegmentId <= latestSegment; currSegmentId++) {
+                    // Close segments before deleting their corresponding log files
+                    String segmentFilePath;
+                    try (Segment sh = getSegmentHandleForSegment(currSegmentId)) {
+                        openSegments.remove(sh.id);
+                        segmentFilePath = sh.segmentFilePath;
+                    }
+                    deleteFilesMatchingFilter(file -> file.getAbsolutePath().equals(segmentFilePath));
+                }
+                if (committedTailSegment > 0) {
+                    newTailAddress = getLastAddressInSegment(committedTailSegment - 1);
+                }
 
-            long newTailAddress = StreamLogDataStore.ZERO_ADDRESS;
-            if (committedTailSegment > 0) {
-                newTailAddress = committedTailSegment * RECORDS_PER_LOG_FILE + 1;
+            } else {
+                // Do not need to delete any segment
+                newTailAddress = globalTail;
             }
 
             logMetadata = new LogMetadata(dataStore);
             logMetadata.syncTailSegment(newTailAddress);
-            initializeLogMetadata();
+            initializeLogMetadata(true);
```

Comment:
This looks like a bug. What would happen in the following scenario:
1. logMetadata is ahead of the committed tail
2. The node leaves and rejoins
3. Node reset removes all the segments [segment(committed tail), last segment]
4. `initializeLogMetadata` is initialized from the Datastore
5. There is a gap between segment(committed tail) and tails initialized from initializeLogMetadata
6. State transfer starts transferring address from segment(committed tail), but the writes dont fail on overwrites because the detection is based on the index from `Segment`
7. Now the `LogMetadata` has addresses that are not consistent with the addresses in `Segment`
8. Heal completes
9. Another node leaves and the sequencer state is bootstrapped from corrupted memory addresses from step 7

Reply:
The delta transfer calls LU reset which sets a new tail address (if committed tail is in segment N, the starting address will be the last address in segment N-1 ).
https://github.com/CorfuDB/CorfuDB/blob/e399b39d61738760a22027c39e84ff10ce2a5ea9/infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java#L636 (btw there is a bug in computing the new tail address)

The initializeLogMetadata will call loadPersistedLogMetadata(maxAddress) where maxAddress is the new tail address. State transfer will transfer from segment N and the address space of those data will be updated in normal log append path.